src/main/java/com/aws/greengrass/shadowmanager/sync/model/MergedFullShadowSyncRequest.java

```diff
@@ -0,0 +1,143 @@
+ */
+
+package com.aws.greengrass.shadowmanager.sync.model;
+
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.shadowmanager.exception.RetryableException;
+import com.aws.greengrass.shadowmanager.exception.SkipSyncRequestException;
+import com.aws.greengrass.shadowmanager.exception.UnknownShadowException;
+import com.aws.greengrass.shadowmanager.sync.RequestMerger;
+import lombok.Getter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.aws.greengrass.shadowmanager.model.Constants.LOG_SHADOW_NAME_KEY;
+import static com.aws.greengrass.shadowmanager.model.Constants.LOG_THING_NAME_KEY;
+
+/**
+ * A {@link FullShadowSyncRequest} that was created by merging other {@link SyncRequest}s
+ * together. This class keeps track of all such merged requests.
+ *
+ * <p>When this sync request is executed, it will look through the merged requests
+ * and determine if a full sync is truly needed, based on {@link BaseSyncRequest#isUpdateNecessary(SyncContext)}.
+ */
+public class MergedFullShadowSyncRequest extends FullShadowSyncRequest {
+
+    private static final Logger logger = LogManager.getLogger(MergedFullShadowSyncRequest.class);
+
+    /**
+     * List of individual {@link SyncRequest}s that were merged together to become a full sync request.
+     */
+    @Getter
+    private final List<SyncRequest> mergedRequests = new ArrayList<>();
+    private final RequestMerger merger;
+
+    /**
+     * Creates a new MergedFullShadowSyncRequest.
+     *
+     * @param thingName      thing name
+     * @param shadowName     shadow name
+     * @param merger         merger
+     * @param mergedRequests requests that were merged to become a full sync request
+     */
+    public MergedFullShadowSyncRequest(String thingName, String shadowName,
+                                       RequestMerger merger, SyncRequest... mergedRequests) {
+        super(thingName, shadowName);
+        this.merger = Objects.requireNonNull(merger);
+        if (mergedRequests != null && mergedRequests.length > 0) {
+            this.mergedRequests.addAll(flatten(Arrays.asList(mergedRequests)));
+        }
+    }
+
+    @Override
+    public void execute(SyncContext context)
+            throws RetryableException, SkipSyncRequestException, InterruptedException, UnknownShadowException {
+        super.setContext(context);
+
+        if (getMergedRequests().stream().anyMatch(r -> r.getClass().equals(FullShadowSyncRequest.class))) {
+            // If a full sync has been directly requested, not due to merging of requests, respect it.
+            super.execute(context);
+            return;
+        }
+
+        List<SyncRequest> necessaryMergedUpdates = getNecessaryMergedRequests(context);
+        if (necessaryMergedUpdates.isEmpty()) {
```

Comment:
We could remove a bunch of ifs if we want to. All this if can go away given `getNecessaryMergedRequests` always will return a non null value and the code below can execute.

Reply:
which ifs?  getNecessaryMergedRequests does always return nonnull

Reply:
I meant you could remove this if and if it is empty it wouldn't process anything when combined with the suggestion below. You would always do `getNecessaryMergedRequests(context).stream()` and would only do things if non empty then this if and the one below would go away.